### PR TITLE
[4249] Added sublabel when maxcharacters bigger than 0

### LINKED ISSUE #4249: 'Maximum X characters' absent under text field or area on PDP if 'Max Characters' is set in admin
**Description**:
'Maximum X characters' absent under text field or area on PDP if parameter 'Max Characters' is set in admin

**Expected behavior**:
'Maximum X characters' present under text field or area on PDP if 'Max Characters' is set in admin
image

**Screenshots, Video, Logs**
image

**To Reproduce**:
1. Open https://master.d25kd40decjpue.amplifyapp.com/default/kids/jeans-for-girls.html

**Actual result**:
'Maximum X characters' absent under text field or area on PDP if 'Max Characters' is set in admin

### PR DESCRIPTION
**Pull request by pablobentivengo**
**Related issue(s):**
* Fixes #4249 

**Problem:**
* 'Maximum X characters' absent under text field or area on PDP if parameter 'Max Characters' is set in admin

**In this PR:**
* Added sublabel when max_characters bigger than 0

### CODE REVIEW

**Comment by aleksandrakorolova** [comment]
1. Text 'Maximum X characters (X remaining)' is bold
Screenshot 2022-02-11 at 09 33 36
**Expected result:**
Text isn't bold

2. Text '(X remaining)' is displayed if user didn't enter text in field
Screenshot 2022-02-11 at 09 45 42

**Expected result:**
Text '(X remaining)' appears when starting to enter text in the field as it works in default Magenta
Screenshot 2022-02-11 at 09 54 31

3. Error message 'Maximum X characters' appears below field if user entered more that max characters
Screenshot 2022-02-11 at 09 50 01
**Expected result:**
Error message 'Maximum X characters (Y too many)' appears as in default Magenta
Screenshot 2022-02-11 at 09 53 30

**Comment by aleksandrakorolova** [comment]
1. Fixed
2. Fixed
3. Please, delete '!' symbol in the error message
4. Text 'Field required' is displayed below field for the required field if user didn't enter text in the field
Screenshot 2022-02-14 at 07 59 10

**Comment by aleksandrakorolova** [comment]
3. Still, can reproduce
4. Still can reproduce but now Text 'Field required' is displayed in the separate line
Screenshot 2022-02-15 at 07 38 12

**Comment by aleksandrakorolova** [comment]
3. Fixed
4. Fixed
5. Different errors appear to products with customizable products when entering more than max characters and trying to add product to cart in condition that all required options are selected

**Product 1:**
Error message 'Incorrect or missing options!' appears
Screenshot 2022-02-16 at 07 49 13 (2)
**Product 2:**
Error message 'The text is too long. Shorten the text and try again.' appears
Screenshot 2022-02-16 at 07 49 07 (2)

**Expected result:**
Error message 'Please enter no more than X characters.' appears below text field/area as it works in default Magenta

**Comment by aleksandrakorolova** [comment]
5. Still, error message 'The text is too long. Shorten the text and try again.' appears when adding product to cart
Screenshot 2022-02-21 at 11 47 44 (2)

6. Error message 'Incorrect or missing options!' is missing when adding config/bundle product to cart (or product with required options)
https://drive.google.com/file/d/11Pc1-99qFjc4Q05hmaJcEAO5WhYzkL0O/view?usp=sharing

**Comment by aleksandrakorolova** [comment]
5. Fixed
6. Fixed
7. Nothing happens when adding config product to cart from PLP if user can select all attributes from PLP
https://drive.google.com/file/d/1ehwTRfMeT5KWH46BuT3TB3IlDLNPfOwh/view?usp=sharing
**Expected result:**
- Shaking effect is displayed and error message appears if user can add product from PLP
- User redirects to PDP if user can't select all attributes from PLP (now it works correctly)

**Comment by aleksandrakorolova** [comment]
7. Fixed
8. Error message 'Please enter no more than X characters.' disappears when adding more characters in the field 
**Steps:**
Open product - https://pr-4275.d25kd40decjpue.amplifyapp.com/default/testing-simple.html
Enter more X symbols in the field
Click on 'Add to cart'
Add one more symbol in the field

**Actual result:**
 Error message 'Please enter no more than X characters.' disappears 
https://drive.google.com/file/d/1l-jUsVttIz0elQ_p3HxcsRUWNdKF7s1s/view?usp=sharing

**Expected result:**
 Error message 'Please enter no more than X characters.' stays while characters won't be equal or less than X

**Comment by aleksandrakorolova** [comment]
8. The expected result is not achieved. Now error message doesn't disappear if user removed extra characters 
https://drive.google.com/file/d/1dICS1QT1s1VxvVl6SL7UNZmQb8SdRhfj/view?usp=sharing

**Comment by pablobentivengo** [comment]
Hey! Right now its working such as in Luma.
https://vimeo.com/683221476/7d34b50847

**Comment by aleksandrakorolova** [comment]
Hey, we have different logic with error messages below field. Please, do the same as written in expected result because all other errors messages work as written in expected result (you can check message 'This field is required')
https://drive.google.com/file/d/1KYq3lpbPzb2O3hDwELJxpJf0R4l8Ljjn/view?usp=sharing

**Comment by aleksandrakorolova** [comment]
8. Please, show this error message 'Please enter no more than X characters.' only when validating form (add product to cart)

**Steps:**
1. Open https://pr-4275.d25kd40decjpue.amplifyapp.com/default/jeans-for-girls.html
2. Enter more than X characters in the field
3. Add product to cart
4. Enter less than X characters in the field
5. Enter more than X characters in the field

**Actual result:**
2. Error message 'Please enter no more than X characters.' is missing (ok)
4. Error message 'Please enter no more than X characters.' is displayed (ok)
5. Error message 'Please enter no more than X characters.' is displayed

**Expected result:**
5. Error message 'Please enter no more than X characters.' is missing 
Error message 'Please enter no more than X characters.' appears only if user entered more than max characters and clicked 'Add to cart'

**Comment by aleksandrakorolova** [comment]
8. Can reproduce: Error message 'Please enter no more than X characters.' disappears when adding more characters in the field
**Steps:**
1. Open product - https://pr-4275.d25kd40decjpue.amplifyapp.com/default/testing-simple.html
2. Enter more X symbols in the field
3. Click on 'Add to cart'
4. Add one more symbol in the field

**Actual result:**
Error message 'Please enter no more than X characters.' disappears
https://drive.google.com/file/d/1oQ_0YsATjhD1wEpIQk8NOt-gZ_xiFkx7/view?usp=sharing

**Expected result:**
Error message 'Please enter no more than X characters.' stays while characters won't be equal or less than X

**Comment by aleksandrakorolova** [comment]
8. Again can reproduce https://github.com/scandipwa/scandipwa/pull/4275#issuecomment-1058072406

File: packages/scandipwa/src/util/Validator/Validator.js [hunk 1]
```diff
@@ -84,7 +84,12 @@ export const validate = (value, rule) => {
             }
 
             if (max && value.length > max) {
-                output.errorMessages.push(onRangeFailMax || __('Maximum %s characters!', max));
+                const tooMany = value.length - max;
```
Comment (AleksandrsKondratjevs) [inline packages/scandipwa/src/util/Validator/Validator.js:87]: Usually we use prefix 'is' for booleans

Comment (AleksandrsKondratjevs) [inline packages/scandipwa/src/util/Validator/Validator.js:87]: Usually we use prefix 'is' for booleans

File: packages/scandipwa/src/component/Product/Product.container.js [hunk 2]
```diff
@@ -405,13 +410,36 @@ export class ProductContainer extends PureComponent {
...
+        if (
+            errorFields
+            || errorMessages
+            || this.validateConfigurableProduct()
+            || this.filterAddToCartFileErrors(values)
+        ) {
+            this.scrollOptionsIntoView();
+            this.setState({ addToCartTriggeredWithError: true });
+            showError(__('Incorrect or missing options!'));
+            return true;
```
Comment (AleksandrsKondratjevs) [inline packages/scandipwa/src/component/Product/Product.container.js:430]: space before return.

Comment (AleksandrsKondratjevs) [inline packages/scandipwa/src/component/Product/Product.container.js:430]: space before return.

**Review by AleksandrsKondratjevs — CHANGES_REQUESTED** [review]

File: packages/scandipwa/src/component/Field/Field.container.js [hunk 3]
```diff
@@ -132,10 +150,22 @@ export class FieldContainer extends PureComponent {
...
         }
-        this.setState({ validationResponse: output });
 
+        // When submit and response equals true (it can be object) reset show length error
+        if (response === true) {
+            this.setState({ showLengthError: false });
+        }
+
+        this.setState({ validationResponse: output });
         return output;
```
[the remark this answers is not in the document]
Comment (pablobentivengo) [reply packages/scandipwa/src/component/Field/Field.container.js:169]: done

[the remark this answers is not in the document]
Comment (pablobentivengo) [reply packages/scandipwa/src/component/Field/Field.container.js:169]: done

File: packages/scandipwa/src/component/ProductCustomizableOption/ProductCustomizableOption.component.js [hunk 4]
```diff
@@ -20,6 +20,8 @@ import FieldGroup from 'Component/FieldGroup';
 import { CustomizableOptionsType } from 'Type/ProductList.type';
 import { customizableOptionToLabel } from 'Util/Product/Transform';
 
+// eslint-disable-next-line no-unused-vars
+import { getSubLabelFromMaxCharacters } from '../../util/Product/Extract';
```
[the remark this answers is not in the document]
Comment (pablobentivengo) [reply packages/scandipwa/src/component/ProductCustomizableOption/ProductCustomizableOption.component.js:24]: done

[the remark this answers is not in the document]
Comment (pablobentivengo) [reply packages/scandipwa/src/component/ProductCustomizableOption/ProductCustomizableOption.component.js:24]: done

File: packages/scandipwa/src/component/ProductCustomizableOption/ProductCustomizableOption.container.js [hunk 5]
```diff
@@ -24,6 +24,7 @@ import { CONFIG_FIELD_TYPE } from './ProductCustomizableOption.config';
 /** @namespace Component/ProductCustomizableOption/Container/mapStateToProps */
 export const mapStateToProps = (state) => ({
     currencyCode: state.ConfigReducer.currencyData.current_currency_code
+
```
[the remark this answers is not in the document]
Comment (pablobentivengo) [reply packages/scandipwa/src/component/ProductCustomizableOption/ProductCustomizableOption.container.js:27]: done

[the remark this answers is not in the document]
Comment (pablobentivengo) [reply packages/scandipwa/src/component/ProductCustomizableOption/ProductCustomizableOption.container.js:27]: done

File: packages/scandipwa/src/util/Product/Extract.js [hunk 6]
```diff
@@ -479,6 +479,33 @@ export const getAdjustedPrice = (product, downloadableLinks, enteredOptions, sel
     return adjustedPrice;
 };
 
+//#region SubLabel
```
[the remark this answers is not in the document]
Comment (pablobentivengo) [reply packages/scandipwa/src/util/Product/Extract.js:482]: done

[the remark this answers is not in the document]
Comment (pablobentivengo) [reply packages/scandipwa/src/util/Product/Extract.js:482]: done

File: packages/scandipwa/src/util/Product/Extract.js [hunk 7]
```diff
@@ -479,6 +479,33 @@ export const getAdjustedPrice = (product, downloadableLinks, enteredOptions, sel
...
+/**
+ * Returns ProductCustomizable Text/TextArea Option Sublabel based on Maximum characters setup in BE
+ * @param maxCharacters
+ * @param value
+ * @namespace Util/Product/Extract/getSubLabelFromMaxCharacters
+ */
+
```
[the remark this answers is not in the document]
Comment (pablobentivengo) [reply packages/scandipwa/src/util/Product/Extract.js:489]: done

[the remark this answers is not in the document]
Comment (pablobentivengo) [reply packages/scandipwa/src/util/Product/Extract.js:489]: done

File: packages/scandipwa/src/component/Product/Product.container.js (hunk 2 above)
Comment (pablobentivengo) [reply packages/scandipwa/src/component/Product/Product.container.js:430]: done

Comment (pablobentivengo) [reply packages/scandipwa/src/component/Product/Product.container.js:430]: done

File: packages/scandipwa/src/component/Field/Field.container.js [hunk 8]
```diff
@@ -117,12 +119,28 @@ export class FieldContainer extends PureComponent {
...
-        const { validationRule, type, attr: { name } = {} } = this.props;
+        const {
+            validationRule: { range }, type, attr: { name } = {}
+        } = this.props;
+        const { showLengthError } = this.state;
         const value = type === FIELD_TYPE.checkbox || type === FIELD_TYPE.radio
             ? !!this.fieldRef.checked
             : this.fieldRef.value;
-        const response = validate(value, validationRule);
+        const maxValidLength = range ? range.max : 0;
```
[the remark this answers is not in the document]
Comment (pablobentivengo) [reply packages/scandipwa/src/component/Field/Field.container.js:141]: But if validationRule doesn't have range defined how should I destruct? something like this?
`const {
            validationRule: { range: { max = 0 } = {} } = this.props;`

File: packages/scandipwa/src/component/Field/Field.container.js [hunk 9]
```diff
@@ -132,10 +150,22 @@ export class FieldContainer extends PureComponent {
...
+                this.setState({ showLengthError: true });
+                output.errorMessages.unshift(__('Please enter no more than %s characters.', maxValidLength));
+            }
+
             data.detail.errors.push(output);
         }
-        this.setState({ validationResponse: output });
 
+        // When submit and response equals true (it can be object) reset show length error
+        if (response === true) {
```
[the remark this answers is not in the document]
Comment (pablobentivengo) [reply packages/scandipwa/src/component/Field/Field.container.js:164]: Tested and it doesn't work

[the remark this answers is not in the document]
Comment (pablobentivengo) [reply packages/scandipwa/src/component/Field/Field.container.js:164]: Tested and it doesn't work

File: packages/scandipwa/src/util/Validator/Validator.js (hunk 1 above)
Comment (pablobentivengo) [reply packages/scandipwa/src/util/Validator/Validator.js:87]: Hey! But tooMany is an Int.

File: packages/scandipwa/src/store/Product/Product.reducer.js [hunk 10]
```diff
@@ -49,7 +50,6 @@ export const ProductReducer = (
             ...state,
             product: getIndexedProduct(product)
         };
-
```
[the remark this answers is not in the document]
Comment (pablobentivengo) [reply packages/scandipwa/src/store/Product/Product.reducer.js:52]: Hey, this file is as it was, I modified it to add an extra action but removed it and left it as is.

Before I applied that extra action it wasn't even exported in the index.js file of Stores folder. (So I removed it after the mods I made)

File: packages/scandipwa/src/component/Field/Field.container.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (AleksandrsKondratjevs) [reply packages/scandipwa/src/component/Field/Field.container.js:141]: yeap

**Review by AleksandrsKondratjevs — COMMENTED** [review]
Resolve

**Pull request merged**